### ISSUE TITLE
Segfault when running a thin shell model on multiple cores.

### ISSUE THREAD

**Issue by Higginbottom** [sirocco-rt/sirocco#870]
I've been running simple thin shell models, but discovered that this .pf file

[input.pf.txt](https://github.com/agnwinds/python/files/6577190/input.pf.txt)

crashes with a segfault on the current dev version

I run it with a command like

mpirun -n 4 py85g input

and it crashes, sometimes with segfaults, sometimes with other odd errors.

this version works:
commit 47cccd73fc7f13a60052d76aea3d05652843a0dc (HEAD)
Author: Knox S. Long <long@stsci.edu>
Date:   Tue Apr 20 14:04:26 2021 -0500

This fails
commit 0d5b2cd3642a3d0e5ed910b16ee5658861f42b95 (HEAD)
Author: Edward Parkinson <e.j.parkinson@soton.ac.uk>
Date:   Fri Apr 30 13:33:35 2021 +0300

**Comment by Higginbottom** [comment]
This 

 06963a2b24c6b7a9703961bdfc4d412638a6854d

seems to work on excession..

but it crashes on my Mac with that particular file.

**Comment by Higginbottom** [comment]
OK, so if the brenstrahlung spectrum temperature is set to 1.2e7 - it runs serial and MPI

Set it to 5.2e7 - which I what I want - runs serial but fails MPI...
5.1e7 - works. What the???

Increasing the maximum frequency of photons produced stops the crash - so its something about not making photons in the exponential tail. But why would it crash?

**Comment by kslong** [comment]
Very odd.  Does it crash with a single core, which would make it easy to find where the seg fault is.

**Comment by Higginbottom** [comment]
Its fine with a single core.... And actually the fundamental issue isnt a seg fault -

This file 
[fails.pf.txt](https://github.com/agnwinds/python/files/6577409/fails.pf.txt)
crashes with 

Zeroth thread successfully received the normalised estimators. About to broadcast.
Zeroth thread successfully received the integer sum. About to broadcast.
No need to communicate matom estimators as no macro-atoms!
!!python: Number of ionizing photons 2.6297e+48 lum of ionizing photons 1.98928e+39
sane_check: 0 nan
normalise_simple_estimators:sane_check 0 ave_freq nan j nan ntot 1000000
sane_check: 0 nan
normalise_simple_estimators:sane_check 0 trad      nan w      nan
normalise_simple_estimators: ave_freq      nan j      nan

but it runs fine in single processor mode. So it looks like there is something going on in the communication of simple estimators. Which could cause a seg fault in some circumstances I guess...

Will see if I can track down any changes to that code between 44cccd7 and 0d5b2cd!

**Comment by Higginbottom** [comment]
working thruogh commits - and running each time - sometimes I get 


[uos-15962:03856] *** Process received signal ***
[uos-15962:03856] Signal: Segmentation fault: 11 (11)
[uos-15962:03856] Signal code:  (0)
[uos-15962:03856] Failing at address: 0x0
[uos-15962:03856] [ 0] 0   libsystem_platform.dylib            0x00007fff7032d5fd _sigtramp + 29
[uos-15962:03856] [ 1] 0   ???                                 0xffffffffffffffff 0x0 + 18446744073709551615
[uos-15962:03856] [ 2] 0   py85g                               0x000000010d8be1eb translate_in_wind + 395
[uos-15962:03856] [ 3] 0   py85g                               0x000000010d8bde78 translate + 88
[uos-15962:03856] [ 4] 0   py85g                               0x000000010d8c9c91 trans_phot_single + 225
[uos-15962:03856] [ 5] 0   py85g                               0x000000010d8c9abe trans_phot + 526
[uos-15962:03856] [ 6] 0   py85g                               0x000000010d91e553 calculate_ionization + 1475
[uos-15962:03856] [ 7] 0   py85g                               0x000000010d8b58fe main + 2766
[uos-15962:03856] [ 8] 0   libdyld.dylib                       0x00007fff70134cc9 start + 1
[uos-15962:03856] [ 9] 0   ???                                 0x0000000000000002 0x0 + 2

**Comment by Higginbottom** [comment]
Annoyingly, it seems that it was my changes implemented to fix problems with the shape of the spectra 

commit 3956c7c26909eec283b26d10deded007dad438ed (HEAD)
Author: higginbottom <nick_higginbottom@fastmail.fm>
Date:   Wed Apr 28 10:44:24 2021 +0100

    Fixes for bug #279

that introduced the problem.
Ugh!!

**Comment by Higginbottom** [comment]
Confusingly, it is simply changing 

FUNC_CDF from 200 to 2000 that is the cause of the problem....

Setting it to 1999 makes the code run fine.

Hmmmmm

I suspect it is an array that is sized incorrectly - since extending the photon range fixes it too - I suspect it is something to do with adding extra elements to an array till we hit a total flux - normally we never make it to the end, coz we are in an exponential tail, but because I have not extended the photon creation into the exp tail - I do add all the elements...

Checking..

**Comment by kslong** [comment]
@Higginbottom    Is there an update on this?

**Comment by kslong** [comment]
OK,  I confirm this problem still exists.  It is fairly bizarre since here is where in python.h the cdf structure is defined

````
#define NCDF 30000              //The default size for these arrays.  This needs to be greater than
                                //the size of any model that is read in, hence larger than NWAVE_EXTRACT in models.h
#define FUNC_CDF  2000          //The size for CDFs made from functional form CDFs
#define ARRAY_PDF 1000          //The size for PDFs to be turned into CDFs from arrays


typedef struct Cdf
{
  double x[NCDF];               /* Positions for which the CDF is calculated */
  double y[NCDF];               /* The value of the CDF at x */
  double d[NCDF];               /* 57i -- the rate of change of the CDF at x */
  double limit1, limit2;        /* Limits (running from 0 to 1) that define a portion
                                   of the CDF to sample */
  double x1, x2;                /* limits if they exist on what is returned */
  double norm;                  /* The scaling factor which would renormalize the CDF */
  int ncdf;                     /* Size of this CDF */
}
 *CdfPtr, cdf_dummy;

````

It's hard to imagine why changing FUNC_CDF from 1999 to 2000 would matter here, as the CDF arrays themselves are huge compared to NCDF.  

Note that for ksl this does not always fail with mpirun, but it does with certain numbers of threads.  For example it fails for 4 threads.  I actually could not find any other set for which it failed.  I don't know if it is relevant but this is exactly the same number of cells in this model.

But what could any of this have to do with FUNC_CDF.

**Comment by kslong** [comment]
Just before the failure, we get
````
estimators_simple.c:477:      Error ("normalise_simple_estimators:sane_check %d ave_freq %e j %e ntot %d\n", xplasma->nplasma, wtest, xplasma->j, xplasma->ntot);
estimators_simple.c:491:      Error ("normalise_simple_estimators: Huge w %8.2e in cell %d trad %10.2e j %8.2e\n", xplasma->w, xplasma->nplasma,
estimators_simple.c:496:      Error ("normalise_simple_estimators:sane_check %d trad %8.2e w %8.2g\n", xplasma->nplasma, radiation_temperature, xplasma->w);
estimators_simple.c:497:      Error ("normalise_simple_estimators: ave_freq %8.2e j %8.2e\n", xplasma->ave_freq, xplasma->j);
````

and all of the values are Nans, except ntot=1000000, and nplasma=0

**Comment by saultyevil** [comment]
You can make this fail in serial mode if you modify the seed to be `init_rand (1084515760 + (13 * 2));` and use 25,000 photons instead of 100000. It'll start spitting errors out in `radiation()`. 

You can track the first sane_check error back to line 363 in radiation.c. Somewhere along the lines `kappa_tot = -NaN`, which makes `tau = -NaN` and then `p->w = -NaN` and so `w_ave_cmf = -NaN` which we use to increment the estimators, so naturally `nplasma->j = -NaN` once this happens.

Think the first NaN appears on line 160 in radiation.c as my debugger says `frac_ff = -NaN`. I think what's happening is we are doing a 0 / 0 in `kappa_ff()` because x1, x2 and x3 are all zero (see below).

![image](https://user-images.githubusercontent.com/19627279/129357820-8095b6a4-ac24-463b-93dd-05528907082b.png)

The electron temperature is 0 in this plasma cell for some reason. This is where I think the -NaN comes in because if x3 was > 0, it would be 0 / number = 0 and not -NaN. I think maybe this could be caused by a gridding issue:

![image](https://user-images.githubusercontent.com/19627279/129357909-07214cf1-cde8-43fe-9fde-c05d6ff5daee.png)

**Comment by kslong** [comment]
@saultyevil   I assumed what I was supposed to do was to substitute the line for init_rand to what you suggested and then run in single processor mode after changing the photon number.  Is that correct.  And then I should be able to produce the error in single processor mode.  Is that correct?  If so, it did not quite seem to work for me.

However, if I look at the last thing you wrote here, it looks like this is happening on a cell, which is not in the wind, presumably our dummy plasma cell.  

So what happens the original way I see the problem with np = 4, is that in thread two, we get the NaNs that are described for a non-existent cell in thread 2, that is the third thread of 4, and then when this is all gathered back in thread 0 in simple_estimators , the fact that the estimators give NaNs causes the program to exit.  Only thread 2, shows the earlier NaNs in radiation.c

**Comment by kslong** [comment]
Indeed, what I just said should have been obvious had we read the logs carefully.  If one looks at the error logs carefully, one see that there was a check_plasma error for thread 2.  Check plasma is called around line 104 of radiation.c, and the call to radiation gets the plasma cell from p->grid  some how it must be very rarely possible to have set p->grid to an incorrect value.

(Note that I think one might argue that the check_plasma error message might be made more explicit!)

**Comment by kslong** [comment]
If you use check_plasma to trap the problem and then not execute the rest of radiation, the crash is avoided and this is what you get for errors:

````
Collated errors for 4 processors
Recurrences --  number of threads with error -- Description
	   28 --     4 -- getatomic_data: line input f odd (may be OK if Macro): %s
	    4 --     4 -- get_atomicdata: Could not interpret line %d in file %s: %s
	    4 --     4 -- Ignored %d inner shell cross sections because no matching yields
	    4 --     4 -- For the shell type wind, we should be in spherical coordinates - changing....
	    4 --     4 -- For the shell type wind, we take control of the grid, and need NDIM to be the minimum - 4 - changing
	    4 --     4 -- wind2d.c: Not currently able to calculate mdot wind for coordtype %d in domain %d
	    3 --     3 -- spherical: ds_in_cell s and smax returning VERY_BIG in cell %i nudging photon by DFUDGE
	    3 --     3 -- translate_in_wind: Grid cell %d of photon is not in wind, moving photon %.2e
	    3 --     3 -- translate_in_wind: photon %d position: x %g y %g z %g
	    3 --     3 -- calculate_ds: the frequency along the photon %d path's in cell %d is the same
	    3 --     3 -- ds_calculate: wind vol = 0 for cell %d photon position %g %g %g
	    1 --     1 -- check_plasma: In Dummy Plasma Cell when probably don't want to be:  %s 
	    1 --     1 -- Radiation: Photon %d is in wind cell %d which has no volume in the wind
````

**Comment by kslong** [comment]
@saultyevil  @Higginbottom I am ready to merge into dev some changes which trap the situation where the routine radiation is called for a photon which is really not in the wind, and then close this issue.  I think this is sufficient as long as this problem  is very rare, but if anyone else wants to pursue exactly what allows this to happen at all, let me know, and I will not close the issue.

**Comment by kslong** [comment]
I have merged the fixes for this model into dev.  The fix traps the situation when radiation is called but the photon says it is in a cell which is not in the actual wind region.  This ultimately caused  one of the threads to exit. 

The program  now throws a much more urgent looking error if the problem occurs again, but it no longer crashes.

Since the error was very rare  and since it only has been seen in the thin shell model, I did not actually determine why the photon thought it was in a wind cell that was not in the wind.
